spec/ic.did

```diff
@@ -44,6 +44,18 @@ service ic : {
   deposit_cycles : (record {canister_id : canister_id}) -> ();
   raw_rand : () -> (blob);
 
+  // Threshold ECDSA signature
+  get_ecdsa_public_key : (record {
+    canister_id : opt canister_id;
+    derivation_path : vec blob;
+    key_id : text;
+  }) -> (record { public_key : blob; chain_code : blob; });
```

Comment:
Can these APIs return error?

Reply:
I think all other APIs are using reject message to deliver errors. So I followed this convention here.

Reply:
As the `chain_code` allows to do arbitrary further derivations, shall we maybe make that an `opt blob` and return `null` unless the request comes from `canister_id = caller` (or `null`).
In particular, if we *always* return the chain code, there would be no need to support a `derivation_path` at all, and all derivation could be done in canister land.

Reply:
The `derivation_path` is here for convenience to avoid having the caller implement derivation by themselves (which would require one set of library for each language we support). The `chain_code` returned is also here for convenience so that we don't force people to do many online calls to calculate a set of derivations.

At least that is my understand, please correct me if I'm wrong @andreacerulli

Reply:
But the `chain_code` would not be public usually. Given the proposed API, any canister can derive keys and get the `chain_code` for any other canister, which is something that, e.g., you couldn't know do for my Bitcoin keys. (Even if I had Bitcoin keys.)

Reply:
This is an interesting question. In the bitcoin world, chain codes can be used to create public keys that are unlinkable. Unlinkable public keys was not something we planned on supporting. One reason is is that to obtain unlinkability, the root chain code has to be secret, and , well the Internet Computer is not very good at keeping secrets, so I'm not this is something we can achieve. Certainly not today. Right now, the root chain code is a hardcoded constant. Anybody with access to our source code can reverse engineer to get the root PK.

Now....we could have a canister supply its own root chain code. Maybe we can't guarantee secrecy now, but maybe someday we can. So we may want to design the API conservatively to allow for this in the future. But then we have to rethink lots of things. As suggested, we could not allow any principal besides the canister than controls the key to obtain the chain code. But to get true unlinkability, I think we would have to go a step further and not allow any principal besides the canister that controls the key to obtain the public key.

We could do that, but I'm not sure if that creates inconveniences and even security problems. To get another canister's PK, we have to ask that canister for it. Not only that, we have to trust the canister to not lie and maybe give us somebody else's public key. We've discussed this before, and while there are no obvious problems with this, I think it could create security footguns...I think people expect to have a reliable way to obtain a public key...but maybe I'm wrong about that. In a typical PKI without PoPs, I can always claim your key is my key, so that is kind of the same.

So I think the larger question is whether only the controlling canister can ask for the public key. If the answer to that is NO, then I think the chain code question is moot. If the answer is YES, then I think the canister can create unlinkable public keys by injecting a secret into the BIPS32 derivation path, and not telling anyone the full derivation path -- I think that is equivalent to secret chain codes.

Let me know if we need a zoom call to discuss.

Reply:
Actually, thinking about this some more, maybe the API as is provides unlinkability. I mean, if a canister wants to have an unlinkable public key, he can generate a subkey using a path that has some secret random element in it, derive the PK, and publish that PK but not the path itself...or at least not the secret random element. That would work. So I think the API is good as is. Also, if we change the API significantly, this would have to be done in concert with Dieter and the bitcoin integration team, as this could impact their designs.

Reply:
Just to summarize the latest offline discussion from the ecdsa weekly meeting, the master ECDSA public keys will be made available in the registry registry, which is public. So anyone would be free to use that information to derive canister public keys given sufficient time of studying our implementation. We tend to agree that trying to hide this from the public interface is only a obscurity which gives a false sense of privacy.

So the conclusion is to keep this interface as is. Any strong objections? @victorshoup-dfinity @Dfinity-Bjoern

Reply:
> Right now, the root chain code is a hardcoded constant. Anybody with access to our source code can reverse engineer to get the root PK.

Since individual chain codes are publicly computable from the master key of the subnet (which can be computed from the registry), I don't see a strong reason to hide them from the API. Maybe we even give a false sense of privacy to the caller if we only return their own chain codes. @Dfinity-Bjoern are you still of the opinion that we should not disclose them?

Reply:
Otherwise, seems fine, so I'll close the thread.